refactor(compiler): allow disabling external symbol factory reexports

devversion opened the pull request:
See individual commits for description of changes.

alexeagle commented:
test failure looks legit?
```
Failures:
1) ngc transformer command-line compile ngfactory files should be able to compile multiple libraries with summaries
  Message:
    Error: Expected lib2_built/module.ngfactory.js to be emitted (outDir: /b/f/w/_tmp/8738305b11d10557227ccf1484986acd/tmp.398296/node_modules)
  Stack:
    Error: Expected lib2_built/module.ngfactory.js to be emitted (outDir: /b/f/w/_tmp/8738305b11d10557227ccf1484986acd/tmp.398296/node_modules)
        at shouldExist (packages/compiler-cli/test/ngc_spec.ts:24:13)
```

alexeagle commented on packages/bazel/src/ng_module.bzl:
```diff
@@ -245,6 +245,10 @@ def _ngc_tsconfig(ctx, files, srcs, **kwargs):
         "fullTemplateTypeCheck": ctx.attr.type_check,
+        # In Blaze we still want to use the symbol factory re-exports in order to
+        # not break existing apps inside google3. Enabling this in google3 would cause
+        # strict dependency failures.
```
@alxhub we also run a strict deps check for Bazel ts_library/ng_module - why doesn't this cause strict deps failures in Bazel?

alxhub replied:
That check only applies to the user's TS code, whereas true StrictDeps in g3 applies to all generated js code as well.

devversion commented:
@alexeagle Yeah test failure looks legit. I'm looking into it.

IgorMinar commented:
is this still WIP?

devversion commented:
@IgorMinar Nope. this is ready now 😄 There are a few things that I'd like to get your opinion about:

1) This is technically a change in behavior since we by default disable symbol re-exports within NGC. So this shouldn't go into patch for now, right?

2) Should I change the `refactor(compiler)` message to `feat(compiler)` because this basically a new flag for the `AotCompiler`, but I wasn't sure if it's worth to be shown in the changelog

3) For NGC we disable symbol re-exports by default, but for `@angular/compiler` I kept the changes backwards-compatible by making the option default to `true`. [see here](https://github.com/angular/angular/pull/28594/files#diff-55c632657b5b6ecc7a41905de326e901R25). It would cleaner to just default to `false` (similar to NGC), but I could imagine that being considered as a breaking change. thoughts?

alexeagle commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:233059242:BASE:233059312:1549640971902:15339df)

alxhub reviewed (approved):
Code-wise, this all looks good to me. To answer your questions.

> 1. This is technically a change in behavior since we by default disable symbol re-exports within NGC. So this shouldn't go into patch for now, right?

I think we should land the change in both master and patch with the default unchanged (always produce re-exports in every case, like we did before) and with the flag removed from parsing, so users can't change it. Then, in a separate PR we can introduce the flag on master only, with default values that disable re-exports in the non-Blaze case.

> 2. Should I change the `refactor(compiler)` message to `feat(compiler)` because this basically a new flag for the `AotCompiler`, but I wasn't sure if it's worth to be shown as feat in the changelog.

No, leave it as `refactor` but remove the publicly visible part of the change (introduction of the flag). That should go in a separate master-only commit that is a `feat`.

> 3. For NGC we disable symbol re-exports by default now, but for `@angular/compiler` I kept the changes backwards-compatible by making the option default to `true`. [see here (https://github.com/angular/angular/pull/28594/files#diff-55c632657b5b6ecc7a41905de326e901R25). It would cleaner to just default to `false` (similar to NGC), but I could imagine that being considered as a breaking change. thoughts?

As above.

devversion commented:
@alxhub Thanks for answering these. This all sounds reasonable to me. I will make the changes.

devversion committed:
refactor(compiler): allow disabling external symbol factory reexports

Currently external static symbols which are referenced by AOT
compiler generated code, will be re-exported in the corresponding
`.ngfactory` files.

This way of handling the symbol resolution has been introduced in
favor of avoding dynamically generated module dependencies. This
behavior therefore avoids any strict dependency failures.

Read more about a particular scenario here: https://github.com/angular/angular/issues/25644#issuecomment-458354439

Now with `ngtsc`, this behavior has changed since `ngtsc` just
introduces these module dependencies in order to properly reference
the external symbol from its original location (also eliminating the need
for factories). Similarly we should provide a way to use the same
behavior with `ngc` because the downside of using the re-exported symbol
resolution is that user-code transformations (e.g. the `ngInjectableDef`
metadata which is added to the user source code), can resolve external
symbols to previous factory symbol re-exports. This is a critical issue
because it means that the actual JIT code references factory files in order
to access external symbols. This means that the generated output cannot
shipped to NPM without shipping the referenced factory files.

A specific example has been reported here: https://github.com/angular/angular/issues/25644#issue-353554070

devversion commented:
**Caretaker**: The saucelabs job timed out due to flakiness. There is a [presubmit](https://test.corp.google.com/ui#id=OCL:233059242:BASE:233059312:1549640971902:15339df) here, but not sure if it's outdated now.
